storage/innobase/srv/srv0srv.cc

```diff
@@ -2675,6 +2681,77 @@ srv_master_do_shutdown_tasks(
 	return(n_bytes_merged || n_tables_to_drop);
 }
 
+/*********************************************************************//**
+Set temporary tablespace to be encrypted if global variable
+innodb_temp_tablespace_encrypt is TRUE. In case of failure, changes
+value of innodb_temp_tablespace_encrypt back to FALSE and logs error
+message. */
+static
+void
+srv_enable_temp_encryption_if_set()
+{
+	if (srv_shutdown_state != SRV_SHUTDOWN_NONE) {
+		return;
+	}
+
+	if (!srv_tmp_tablespace_encrypt) {
+		return;
+	}
+
+	if (is_shared_system_tablespace(srv_tmp_space.space_id())) {
+		/* there is no separate tablespace for temporary tables */
+		return;
+	}
+
+	fil_space_t* space = fil_space_get(srv_tmp_space.space_id());
+
+	ut_ad(fsp_is_system_temporary(space->id));
+
+	if (!FSP_FLAGS_GET_ENCRYPTION(space->flags)) {
+
+		if (srv_read_only_mode) {
+			srv_tmp_tablespace_encrypt = false;
+			ib::error() << "Can't set temporary tablespace "
+				    << "to be encrypted in InnoDB read-only "
+				    << "mode.";
+			return;
+		}
+
+		/* Make sure the keyring is loaded. */
+		if (!Encryption::check_keyring()) {
+			srv_tmp_tablespace_encrypt = false;
+			ib::error() << "Can't set temporary tablespace "
+				    << "to be encrypted because "
+				    << "keyring plugin is not "
+				    << "available.";
+			return;
+		}
+
+		space->flags |= FSP_FLAGS_MASK_ENCRYPTION;
+
+		dberr_t	err = fil_set_encryption(space->id,
+			Encryption::AES, NULL, NULL);
+
+		ut_a(err == DB_SUCCESS);
+
+#ifdef UNIV_DEBUG
+		if (srv_master_encrypt_debug != 0) {
+			srv_master_encrypt_debug = 2;
```

Comment:
How is this working without atomics? Not even volatile, which would also be wrong, but at least it would force loads and stores as written

Reply:
There is a `os_thread_sleep` call in the loop which can potentially use and modify any global variable, compiler would be terribly wrong if it would optimize `srv_master_encrypt_debug = 2` away.